lib/SIL/SILFunctionType.cpp

```diff
@@ -2213,12 +2225,27 @@ namespace {
         substParams.push_back(subst(origParam));
       }
 
-      return SILFunctionType::get(nullptr,
-                                  origType->getExtInfo(),
-                                  origType->getCalleeConvention(),
-                                  substParams, substResults,
-                                  substErrorResult,
-                                  getASTContext());
+      Optional<ProtocolConformanceRef> witnessMethodConformance;
+      if (auto conformance = origType->getWitnessMethodConformanceOrNone()) {
+        assert(origType->getExtInfo().hasSelfParam());
+        auto selfType = origType->getSelfParameter().getType();
+        // The Self type can be nested in a few layers of metatypes (etc.), e.g.
```

Comment:
Example:

```swift
protocol Foo {
  static var x: Int { get set }
}

func mutate(_: inout Int) {}
func foo<T: Foo>(ty: T.Type) {
  mutate(&T.x)
}

```

The SIL for the setter callback call after calling materializeForSet looks like:

```
...
bb2(%16 : $Builtin.RawPointer):                   // Preds: bb0
  %17 = pointer_to_thin_function %16 : $Builtin.RawPointer to $@convention(witness_method: Foo) <τ_0_0 where τ_0_0 : Foo> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @inout @thick τ_0_0.Type, @thick τ_0_0.Type.Type) -> () // user: %22
...
```